main/src/library/ChannelImpl.flix

```diff
@@ -0,0 +1,374 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+namespace ChannelImpl {
+
+    ///
+    /// The tuple corresponds to (
+    ///   `id` - an unique id used to order the locking of sets of channels. This avoids deadlocks,
+    ///   `channelLock` - the lock of the channel,
+    ///   `buffered` - indicates whether a handoff is needed or not.
+    ///       If you try to put an element in a channel that's full, you wait until there's space,
+    ///   `maxQueueSize` - the maximum number of elements allowed in `deque`. This is `max(1, channel size)`,
+    ///   `deque` - a deque of the channel elements,
+    ///   `waitingGetters` - a set of conditions that are waiting for an element.
+    ///       This set is cleared after each new element,
+    ///   `waitingSetters` - a condition that can notify threads of available space in `deque`.
+    /// )
+    /// The reason for the difference between `waitingGetters` and `waitingSetters` is the select expression.
+    /// A thread can only wait for one condition which is a problem when a select expression requires waiting for a
+    /// set of channels. Therefore the thread makes a condition that represents the select expression and hands that to
+    /// every channel involved.
+    ///
+    /// put is not supported in select so a single condition can be reused for all insertions.
+    ///
+    pub enum ChannelImpl {
+        case Mpmc(
+                Int64,
+                ReentrantLock,
+                Bool,
+                Int32,
+                MutDeque[Boxed],
+                MutList[(ReentrantLock, Condition)],
+                Condition
+            )
+    }
+
+    ///
+    /// Equality is based on the channel `id` and is needed for `Order`.
+    ///
+    instance Eq[ChannelImpl] {
+      pub def eq(x: ChannelImpl, y: ChannelImpl): Bool =
+        let Mpmc(a_id, _, _, _, _, _, _) = x;
+        let Mpmc(b_id, _, _, _, _, _, _) = y;
+        a_id == b_id
+    }
+
+
+    ///
+    /// Order is based on the channel `id` and is used by `selectImpl` to lock in a global ordering.
+    ///
+    instance Order[ChannelImpl] {
+      pub def compare(x: ChannelImpl, y: ChannelImpl): Comparison =
+        let Mpmc(a_id, _, _, _, _, _, _) = x;
+        let Mpmc(b_id, _, _, _, _, _, _) = y;
+        a_id <=> b_id
+    }
+
+
+    ///
+    /// Creates a new channel. A runtime error occurs if `bufferSize` is negative.
+    ///
+    pub def new(bufferSize: Int32): ChannelImpl & Impure =
+        import dev.flix.runtime.GlobalCounter:newId();
+        let _bufferCheck = if (bufferSize < 0) bug!("bufferSize < 0") else ();
+        let buffered = bufferSize == 0;
+        let reentrantLock = ReentrantLock.new(false);
+        Mpmc(
```

Comment:
Would it make sense to split Mpmc into two constructors instead of having a boolean or would that just lead to too much code duplication? Thoughts?

Reply:
It is only used for the last part of the put method but would require another three lines at every let match on channel. It feels nice to separate them but their behavior only differs on one line.